### ISSUE TITLE
[selectors] `:focusable` pseudo-class

### ISSUE THREAD

**Issue by LeaVerou** [w3c/csswg-drafts#7269]
I was reading [this article](https://www.smashingmagazine.com/2022/04/cta-modal-build-web-component/) and I reached this gem:

```js
const FOCUSABLE_SELECTORS = [
  '[contenteditable]',
  '[tabindex="0"]:not([disabled])',
  'a[href]',
  'audio[controls]',
  'button:not([disabled])',
  'iframe',
  "input:not([disabled]):not([type='hidden'])",
  'select:not([disabled])',
  'summary',
  'textarea:not([disabled])',
  'video[controls]',
].join(',');
```

Authors should not have to do all this just to target focusable elements. Adding a pseudo-class for focusability seems like pretty low-hanging fruit (so low-hanging that I'd swear I've seen it discussed before, but I couldn't find anything so…)

**Comment by jimmyfrasche** [comment]
My (perhaps naive) expectation is that it would mean whatever is necessary so that if I called the `focus` method on a node that matches `:focusable` that element would then become the `document.activeElement`.

**Comment by LeaVerou** [comment]
FWIW, my expectation was similar to what @jimmyfrasche described above.

> It _could_ be a simple alias

It shouldn't. We already have proposals [1](https://css.oddbird.net/toggles/explainer/) [2](https://github.com/openui/open-ui/blob/main/research/src/pages/focusgroup/focusgroup.explainer.mdx) where focusability is affected via CSS properties.

**Comment by jimmyfrasche** [comment]
If it can't include/exclude everything then `:focusable` is a bad name and it should be renamed to something like `:focusable-candidate`.

Related, maybe 90% of my focus woes would be fixed by a javascript api that let me focus the element that the user would land on if they hit tab/shift+tab if focus was on a particular element so I could move focus off my component without having to know about the world outside its perimeter, something like `document.focusNext({from: myComponentRoot})`

**Comment by Loirooriol** [comment]
This seems problematic due to circularities like `:focusable { display: none }`

Also, in Blink `element.focus()` inside `content-visibility: auto` forces style computation in order to determine if the element is focusable. Having `:focusable` do that would defeat `content-visibility: auto`.

So `:focusable-candidate` can be better but it's not clear how it should be defined.

**Comment by emilio** [comment]
Yeah, also in all browsers `visibility: hidden` prevents focusability.

**Comment by emilio** [comment]
(And scrollers are focusable by keyboard iff they are scrollable, so that would be another source of circularity)

**Comment by kizu** [comment]
In my practice, I would probably need to only be the “alias”, so something like `:focusable-candidate` would be very welcome.

It could be defined as something that could be focusable when there are no styles applied, basically, browser would need to only look at the HTML of an element, and don't do anything with styles.

More to that — in my use-cases the `:focusable` would almost always go alongside `:hover` or `:active`, as I often want to style something that is located inside an interactive hovered element (and not just _any_ hovered element). We already can do this for focused elements, but can't for hovered.

**Comment by woody-li** [comment]
What about `tabbable`, the difference is negative tabindex cases.
According to the [specification](https://html.spec.whatwg.org/multipage/interaction.html#focusable), `focusable` includes all `non-null tabindex` values.

The `focusable` almost equal to elements which can be focus by `tab` key.
So I suggest rename as `tabbable`, or both focusable and tabbable.
* `focusable` contains all non-null tabindex.
* `tabbable` only contains non-negative tabindex.

**Comment by LeaVerou** [comment]
That's an interesting idea. Whether it's a good path forwards depends on the use cases for targeting focusable elements. Do we have a list of use cases?

**Comment by jimmyfrasche** [comment]
If this existed one thing I'd use it for would be quick debugging css (`:focusable{outline:3px dotted hotpink}`) but the main thing I'd want to do is find the next `:focusable` in js and focus it for when I need to manage focus exiting a complex component (cf my previous comment in this thread)

**Comment by woody-li** [comment]
In some cases, need custom tab navigation focus, so need to know which element is focusable.

A difficult case is the keyboard-focusable scroll containers.
- Firefox has support for it.
- Chrome will support it in a subsequent version.

It's hard to make it compatible with browsers support it or not.

**Comment by andyearnshaw** [comment]
@LeaVerou - https://github.com/w3c/csswg-drafts/issues/7269#issuecomment-2117074783
> That's an interesting idea. Whether it's a good path forwards depends on the use cases for targeting focusable elements. Do we have a list of use cases?

I'm working on a focus trap for widgets on a page. The widgets can be arranged by the user and potentially contain many focusable elements so, in order to help the user quickly navigate around widgets, they can tab/arrow key between them and then press the Enter key to "enter" the active widget and then tab between elements inside it. This is similar to how Microsoft's Power BI web application works (which is also not perfect, btw!):

https://github.com/user-attachments/assets/0a6f4e37-f55b-48fb-b1a0-e3463150edc8

Creating such a trap is definitely tricky. I've seen a lot of poor attempts at a `:tabbable` selector but the biggest problem is that some tababble elements cannot even be selected via CSS. For instance, any element with `overflow: scroll` or `overflow: auto` becomes tababble, but no selector exists to target them based on this state. `display: none` is another property that can affect general focusability without any selector.

I've ended up with something much more complex than I would have liked and resorted to having a way for elements to opt-in to the trap with a data attribute instead of creating a fully-generic, self-contained wrapper function (which beats adding intersection observers to every single descendant). I had to come up with a selector that would target most elements. I've possibly missed some edge cases myself that `:tabbable` might have covered.

All this being said, I don't think either `:tabbable` or `:focusable` are viable selectors due to their potential to be affected by layout, [creating infinite loops](https://wiki.csswg.org/faq#selectors-that-depend-on-layout). Consider the following:

```css
button:focusable {
  display: none;
}
```

`display: none` makes the button no longer focusable, which removes the `display: none`, which makes the button focusable, which reapplies the `display: none`, which... you get the idea. `:overflowed-content` was rejected for [the same reason](https://github.com/w3c/csswg-drafts/issues/2011#issuecomment-774190135).

**Comment by jimmyfrasche** [comment]
That sounds like another case that could be solved by being able to say in javascript "move focus as if the specified element were focused and the user hit tab or shift+tab" then you could just keep track of an element in your component before anything focusable and use that to "shift+tab" out and keep track of an element after everything focusable in your element and use it to "tab" out. That way you don't need to figure out what the previous/next focusable thing is yourself, just leave it up to the browser.
